File: intellinum/intellinum/report/variable_pay_report/variable_pay_report.py
import frappe
from frappe.utils import getdate, today
from datetime import date

MONTH_NAMES = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]


def execute(filters=None):
    filters = filters or {}
    return get_columns(), get_data(filters)


def get_columns():
    return [
        {"fieldname": "employee", "label": "Employee ID", "fieldtype": "Link", "options": "Employee", "width": 120},
        {"fieldname": "employee_name", "label": "Employee Name", "fieldtype": "Data", "width": 220},
        {"fieldname": "percentage", "label": "Percentage (%)", "fieldtype": "Data", "width": 110},
        {"fieldname": "performance_remark", "label": "Performance Remark", "fieldtype": "Data", "width": 300},
        {"fieldname": "action", "label": "Action", "fieldtype": "Data", "width": 110},
    ]


def get_data(filters):
    if not filters.get("payroll_month"):
        filters["payroll_month"] = MONTH_NAMES[getdate().month - 1]

    conditions = [
        "`tabEmployee`.status = 'Active'",
        "IFNULL(`tabEmployee`.custom_quarter_variable, 0) > 0",
        "`tabEmployee`.custom_variable_pay_months LIKE %(month_like)s",
        "(`tabEmployee`.date_of_joining IS NULL OR NOT (YEAR(`tabEmployee`.date_of_joining) = %(current_year)s AND DATE_FORMAT(`tabEmployee`.date_of_joining, '%%b') = %(payroll_month)s))",
    ]

    user_roles = frappe.get_roles(frappe.session.user)
    if "HR Manager" not in user_roles and "Administrator" not in user_roles:
        logged_in_emp = frappe.db.get_value("Employee", {"user_id": frappe.session.user}, "name")
        if logged_in_emp:
            conditions.append("`tabEmployee`.reports_to = %(reports_to)s")
            filters["reports_to"] = logged_in_emp
        else:
            conditions.append("1 = 0")

    if filters.get("company"):
        conditions.append("`tabEmployee`.company = %(company)s")
    if filters.get("department"):
        conditions.append("`tabEmployee`.department = %(department)s")

    filters["month_like"] = f"%{filters['payroll_month']}%"
    filters["current_year"] = getdate().year

    sql = f"""
        SELECT
            `tabEmployee`.name AS employee,
            `tabEmployee`.employee_name AS employee_name,
            `tabEmployee`.custom_quarter_variable AS quarter_variable,
            vpd.name AS vpd_name,
            vpd.percent AS approved_percent,
            vpd.comments AS performance_remark,
            vpd.variable_given AS variable_given,
            CASE
                WHEN vpd.name IS NOT NULL THEN 'Approved'
                ELSE 'Pending'
            END AS action
        FROM `tabEmployee`
        LEFT JOIN `tabVariable Pay Disbursement` vpd
            ON vpd.name = (
                SELECT vpd2.name
                FROM `tabVariable Pay Disbursement` vpd2
                WHERE vpd2.employee = `tabEmployee`.name
                  AND YEAR(vpd2.payroll_disburse_date) = %(current_year)s
                  AND DATE_FORMAT(vpd2.payroll_disburse_date, '%%b') = %(payroll_month)s
                  AND vpd2.docstatus = 1
                ORDER BY vpd2.creation DESC
                LIMIT 1
            )
        WHERE {" AND ".join(conditions)}
        ORDER BY `tabEmployee`.name
    """

    rows = frappe.db.sql(sql, filters, as_dict=True)

    data = []
    for row in rows:
        item = {
            "employee": row["employee"],
            "employee_name": row["employee_name"],
            "action": row["action"],
        }

        if row.get("vpd_name"):
            percent = row.get("approved_percent")
            if percent is None:
                base = float(row.get("quarter_variable") or 0)
                paid = float(row.get("variable_given") or 0)
                percent = round((paid / base) * 100) if base > 0 else 0

            item["percentage"] = str(int(round(float(percent or 0))))
            item["performance_remark"] = row.get("performance_remark") or "-"
        else:
            item["percentage"] = ""
            item["performance_remark"] = ""

        data.append(item)

    return data


@frappe.whitelist()
def process_single_approval(employee, percentage, performance_remark, company=None, payroll_month=None):
    if not employee:
        frappe.throw("Employee is required.")

    try:
        percentage = float(percentage)
    except Exception:
        frappe.throw("Percentage must be a valid number.")

    if percentage <= 0 or percentage > 100:
        frappe.throw("Percentage must be between 1 and 100.")

    if not performance_remark or not performance_remark.strip():
        frappe.throw("Performance Remark is mandatory.")

    emp = frappe.get_doc("Employee", employee)

    quarter_variable = float(emp.get("custom_quarter_variable") or 0)
    variable_full_year = float(emp.get("custom_variable_amount") or 0)

    if quarter_variable <= 0:
        frappe.throw("Quarter Variable is not configured in Employee master.")

    if variable_full_year <= 0:
        variable_full_year = quarter_variable * 4

    month_name = payroll_month or MONTH_NAMES[getdate().month - 1]

    try:
        month_index = MONTH_NAMES.index(month_name) + 1
    except ValueError:
        frappe.throw(f"Invalid payroll month: {month_name}")

    payroll_date = date(getdate().year, month_index, 1)

    existing_vpd = frappe.db.sql(
        """
        SELECT name
        FROM `tabVariable Pay Disbursement`
        WHERE employee = %s
          AND YEAR(payroll_disburse_date) = %s
          AND MONTH(payroll_disburse_date) = %s
          AND docstatus = 1
        LIMIT 1
        """,
        (employee, payroll_date.year, payroll_date.month),
        as_dict=True,
    )

    if existing_vpd:
        frappe.throw(f"Variable Pay already approved for {emp.employee_name} in {month_name}.")

    vpd = frappe.get_doc({
        "doctype": "Variable Pay Disbursement",
        "employee": employee,
        "employee_name": emp.employee_name,
        "variable_full_year": variable_full_year,
        "quarter_variable": quarter_variable,
        "percent": percentage,
        "comments": performance_remark.strip(),
        "disburse_date": today(),
        "payroll_disburse_date": payroll_date,
    })

    vpd.insert(ignore_permissions=True)
    vpd.submit()

    return {
        "message": "Success",
        "vpd": vpd.name,
    }

File: intellinum/intellinum/report/variable_pay_report/variable_pay_report.js
// Determine if the user has admin-level access for this report
const isAdminOrHR = frappe.user.has_role("Administrator") || frappe.user.has_role("HR Manager");

frappe.query_reports["Variable Pay Report"] = {
    filters: [
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            reqd: 1,
            default: frappe.defaults.get_user_default("Company")
        },
        {
            fieldname: "payroll_month",
            label: __("Payroll Month"),
            fieldtype: "Select", // Changed from Data to Select
            options: "Jan\nFeb\nMar\nApr\nMay\nJun\nJul\nAug\nSep\nOct\nNov\nDec", // Dropdown options
            reqd: 1,
            read_only: isAdminOrHR ? 0 : 1, // Editable for Admins/HR, locked for Reporting Managers
            default: new Date().toLocaleString("en-US", { month: "short" })
        },
        {
            fieldname: "department",
            label: __("Department"),
            fieldtype: "Link",
            options: "Department"
        }
    ],

    get_datatable_options(options) {
        return Object.assign(options, {
            cellHeight: 42
        });
    },

	formatter(value, row, column, data, default_formatter) {
		if (!column) return value;

		if (column.fieldname === "employee" && data?.employee) {
			return `
				<a href="/app/employee/${data.employee}" class="vp-employee-link">
					${frappe.utils.escape_html(data.employee)}
				</a>
			`;
		}

		if (!data?.employee) {
			return default_formatter ? default_formatter(value, row, column, data) : value;
		}

		if (data.action === "Approved") {
			if (column.fieldname === "percentage") {
				return `
					<div class="vp-approved-percent">
						${data.percentage || 0}%
					</div>
				`;
			}

			if (column.fieldname === "performance_remark") {
				return `
					<div
						class="vp-approved-remark"
						title="${frappe.utils.escape_html(data.performance_remark || "")}">
						${frappe.utils.escape_html(data.performance_remark || "-")}
					</div>
				`;
			}

			if (column.fieldname === "action") {
				return `
					<span class="vp-approved-badge">
						✓ Approved
					</span>
				`;
			}
		} else {
			if (column.fieldname === "percentage") {
				return `
					<input
						type="number"
						min="1"
						max="100"
						class="form-control percentage-input vp-percent-input"
						placeholder="%">
				`;
			}

			if (column.fieldname === "performance_remark") {
				return `
					<input
						type="text"
						class="form-control remark-input vp-remark-input"
						placeholder="Performance remark">
				`;
			}

			if (column.fieldname === "action") {
				const emp = (data.employee || "").replace(/'/g, "\\'");
				const empName = (data.employee_name || "").replace(/'/g, "\\'");
				return `
					<button class="btn btn-primary btn-sm vp-approve-btn"
						onclick="window._vpApprove(this, '${emp}', '${empName}')">
						Approve
					</button>
				`;
			}
		}

		return default_formatter ? default_formatter(value, row, column, data) : value;
	},

	onload(report) {
		if (!$("#vp-report-style").length) {
			$("head").append(`
				<style id="vp-report-style">
					.dt-scrollable {
						border-radius: 10px;
						overflow: hidden;
					}

					.dt-cell {
						align-items: center !important;
						padding-top: 0 !important;
						padding-bottom: 0 !important;
					}

					.dt-cell__content {
						width: 100%;
					}

					.vp-employee-link {
						font-weight: 600;
						color: var(--text-color);
						text-decoration: none;
					}

					.vp-employee-link:hover {
						text-decoration: underline;
					}

					.vp-percent-input {
						height: 28px !important;
						width: 90px !important;
						text-align: center;
						font-size: 12px;
						border-radius: 6px;
					}

					.vp-remark-input {
						height: 28px !important;
						font-size: 12px;
						border-radius: 6px;
					}

					.vp-approve-btn {
						height: 28px;
						min-width: 90px;
						font-size: 12px;
						font-weight: 600;
					}

					.vp-approved-badge {
						display: inline-flex;
						align-items: center;
						gap: 4px;
						font-size: 12px;
						font-weight: 600;
						color: #1f9d55;
					}

					.vp-approved-percent {
						font-weight: 600;
						text-align: center;
					}

					.vp-approved-remark {
						color: var(--text-muted);
						overflow: hidden;
						text-overflow: ellipsis;
						white-space: nowrap;
					}

					.dt-row {
						min-height: 42px !important;
					}
				</style>
			`);
		}

		window._vpUpdateRow = function(btn, percentage, remark) {
			const $row = $(btn).closest(".dt-row");

			$row.find(".percentage-input")
				.parent()
				.html(`
					<div class="vp-approved-percent">
						${percentage}%
					</div>
				`);

			$row.find(".remark-input")
				.parent()
				.html(`
					<div class="vp-approved-remark">
						${frappe.utils.escape_html(remark || "-")}
					</div>
				`);

			$(btn)
				.parent()
				.html(`
					<span class="vp-approved-badge">
						✓ Approved
					</span>
				`);
		};

		window._vpApprove = function(btn, employee, empName) {
			const $row = $(btn).closest(".dt-row");
			const percentageVal = $row.find(".percentage-input").val();
			const remarkVal = ($row.find(".remark-input").val() || "").trim();

			if (!percentageVal) {
				frappe.msgprint(__("Please enter a percentage value for {0}.", [empName]));
				return;
			}

			const percentage = parseFloat(percentageVal);
			if (isNaN(percentage) || percentage <= 0 || percentage > 100) {
				frappe.msgprint(__("Percentage must be between 1% and 100%."));
				return;
			}

			if (!remarkVal) {
				frappe.msgprint(__("Please enter a performance remark."));
				return;
			}

			frappe.confirm(
				`Approve <b>${percentage}%</b> Variable Pay for <b>${frappe.utils.escape_html(empName)}</b>?`,
				function() {
					$(btn).prop("disabled", true).text("Approving...");

					frappe.call({
						method: "intellinum.intellinum.report.variable_pay_report.variable_pay_report.process_single_approval",
						args: {
							employee: employee,
							percentage: percentage,
							performance_remark: remarkVal,
							company: frappe.query_report.get_filter_value("company"),
							payroll_month: frappe.query_report.get_filter_value("payroll_month")
						},
						callback(r) {
							if (!r.exc) {
								window._vpUpdateRow(btn, percentage, remarkVal);
								frappe.show_alert({
									message: `${empName} approved successfully`,
									indicator: "green"
								});
								setTimeout(() => {
									frappe.query_report.refresh();
								}, 300);
							} else {
								$(btn).prop("disabled", false).text("Approve");
							}
						},
						error() {
							$(btn).prop("disabled", false).text("Approve");
						}
					});
				},
				function() {
					$(btn).prop("disabled", false).text("Approve");
				}
			);
		};
	}
};